pika/adapters/select_connection.py

```diff
@@ -276,7 +260,9 @@ def process_timeouts(self):
         # were only run periodically.
         for t in sorted(to_run, key=itemgetter('deadline')):
             t['callback']()
-            del self._timeouts[hash(frozenset(t.items()))]
+            k = hash(frozenset(t.items()))
+            if k in self._timeouts:
```

Comment:
@shinji-s: also consider the possibility that one timer callback can unregister additional timers. If timer A callback unregisters timers B and C that were also ready, then B and C callbacks should **not** be called. So you need something like this to really fix the bug:

```
for t in sorted(to_run, key=itemgetter('deadline')):
    k = hash(frozenset(t.items()))
    if k in self._timeouts:
        t['callback']()
        self._timeouts.pop(k, None)
```